### ISSUE TITLE
Set up GitLab/Travis CI

### ISSUE THREAD

**Issue by dzhoshkun** [NifTK/NiftyNet#105]
Related to https://cmiclab.cs.ucl.ac.uk/CMIC/NiftyNet/issues/269

We should set up CPU-only CI tests on Travis until a long-term solution for GPU testing has been found.

**Comment by dzhoshkun** [comment]
* [ ] would the Travis `matrix` work for the different scripts currently available?

**Comment by dzhoshkun** [comment]
https://docs.travis-ci.com/user/languages/python

**Comment by tomdoel** [comment]
For the [ImageSplit](https://github.com/gift-surg/ImageSplit) project, I wanted to avoid different testing paths for travis, gitlab-ci and local testing. So I set up [tox.ini](https://github.com/gift-surg/ImageSplit/blob/master/tox.ini) to do most of the CI steps using virtualenvs, and then configured [.travis.yml](https://github.com/gift-surg/ImageSplit/blob/master/.travis.yml) and [.gitlab-ci.yml](https://github.com/gift-surg/ImageSplit/blob/7fd7680afaaa59b7e7b12bde0bc0fdd27b3b8294/.gitlab-ci.yml#L19-L26) to run the tests using the tox command. Another advantage is I can run any of the CI stages locally; e.g. `tox` will run all the tests; `tox -e lint` will run the linting etc.

**Comment by tomdoel** [comment]
A failing test example with tox: https://cmiclab.cs.ucl.ac.uk/GIFT-Surg/ImageSplit/-/jobs/42111

**Comment by dzhoshkun** [comment]
* [ ] configure tox: sample only!
* [ ] set up Travis to use tox script
* [ ] consider opening a new issue for translating the existing CI scripts to tox (or maybe multiple issues, with one for each part)

**Comment by dzhoshkun** [comment]
* [ ] `testenv:lint` not run

**Comment by dzhoshkun** [comment]
* [ ] revert above commit (tox-lint)

**Comment by dzhoshkun** [comment]
@tomdoel the `lint` testenv does not get run, unless a`travis` section is added (see above commit), due to [the way Travis handles build matrices](https://tox-travis.readthedocs.io/en/stable/envlist.html#advanced-configuration).

**Comment by dzhoshkun** [comment]
* [ ] Even the "quick" tests take too long on Travis, possibly due to the lack of a GPU.

**Comment by dzhoshkun** [comment]
* [ ] [`export` not working on Travis](https://travis-ci.org/NifTK/NiftyNet/jobs/386521748)

**Comment by dzhoshkun** [comment]
* [ ] how to deal with forks, as Travis needs to be set up for each fork separately?

**Comment by dzhoshkun** [comment]
* [ ] rebase onto `dev` after #114 resolved

**Comment by dzhoshkun** [comment]
[`redefined-variable-type` was deprecated already in 2017](https://github.com/PyCQA/pylint/issues/1279#issuecomment-274180145), might have just been removed by now.

**Comment by dzhoshkun** [comment]
`redefined-variable-type` became an extension (unavailable by default) [as of pylint 1.7](https://github.com/PyCQA/pylint/blob/17ed43115509ae76c2768b61136f99de593f91e3/doc/whatsnew/1.7.rst#other-changes)

**Comment by dzhoshkun** [comment]
@wyli As of the above commit, Travis CI is set up and the `pylintjob` and `quicktest` sections of GitLab CI have been translated to Travis CI. You can actually click the red cross next to the commit hash above and see the result for yourself 😄 

There are two issues, though:

* I haven't been able to get pylint to run, mainly [due to version incompatibilities](https://github.com/NifTK/NiftyNet/issues/105#issuecomment-394298742) (current version: 1.9.1, on Adelie we have 1.5.2)
* But the actual problem is that it takes about 30 mins to run even the `quicktest` section, due to the absence of GPU computing on Travis.

**Comment by wyli** [comment]
@dzhoshkun Looks nice! I'll send you a merge request to fix the pylint tests (we could take this opportunity to disucss/test the new workflow for contributors #98 ).

**Comment by dzhoshkun** [comment]
There seems to be [another, (given the current setup) possibly easier, way to continue testing with our custom hardware](https://cmiclab.cs.ucl.ac.uk/CMIC/NiftyNet/issues/269#note_145228) (since GitLab v10.6 on 22 March as far as I can see).

**Comment by dzhoshkun** [comment]
* [ ] finalise the GitLab CI/CD for GitHub test: https://gitlab.com/dzhoshkun/cpp-py-example/services/github/edit (imported from https://github.com/dzhoshkun/cpp-py-example)

**Comment by dzhoshkun** [comment]
@wyli @tomdoel I've managed to [set up GitLab CI for one of my personal projects on GitHub](https://docs.gitlab.com/ee/ci/ci_cd_for_external_repos/github_integration.html#connect-with-personal-access-token), running its CI pipelines on Adelie. For the result, please see [this commit](https://github.com/dzhoshkun/cpp-py-example/issues/27#ref-commit-5e6c991) (click the green tick!) and also [this PR](https://github.com/dzhoshkun/cpp-py-example/pull/28) showing "All checks passed".

For the record, [this feature has been around for a while](https://gitlab.com/gitlab-org/gitlab-ee/issues/3835), but [involved a manual setup, and did not transmit the CI pipeline result back to GitHub](https://gitlab.com/gitlab-org/gitlab-ee/issues/4839).

Also for the record, I've imported the project **not** from CmicLab to GitLab, but from GitHub to GitLab. Mentioning this explicitly, as currently there seems to be an issue exporting projects from CmicLab.

**Comment by wyli** [comment]
Cool, looks like a feasible solution for GPU tests.

**Comment by wyli** [comment]
Sometimes CI runner couldn't find `pylint` command (https://gitlab.com/NifTK/NiftyNet/-/jobs/80703023) any idea?
